### DISCUSSION TITLE
How is database-oriented multi-tenancy implemented?

### DISCUSSION THREAD

**Discussion by jubingc** [milvus-io/milvus#33661] [Q&A and General discussion]
Hi,

I am currently reading the "Database-oriented multi-tenancy" section in the [Multi-tenancy strategies](https://milvus.io/docs/multi_tenancy.md). I have a couple of questions:

1. How are data isolation and search performance ensured for different tenants?
2. Why might resources be wasted on idle tenants?
3. Is there a way to monitor resource usage (CPU/memory/storage/network) by tenants?

Thank you!

**Comment by yhmo** [comment]
Each time you create a new collection/partition, milvus will add more threads to manage the collection/partition. With a larger number of collections/partitions, milvus could run into trouble or slow response. So, it is strongly recommended to keep the number of collections/partitions under 1000(depends on resources(CPU/RAM, high-perf disk), with rich resources, it is better to keep the number under 5000).

Partition key is a good solution for handling millions of users: https://milvus.io/docs/multi_tenancy.md#Partition-oriented-multi-tenancy

For example, we create a collection with a partition key field: "username", is_partition_key=True

```
fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="vector", dtype = DataType.FLOAT_VECTOR, dim=dim),
    FieldSchema(name="username", dtype = DataType.VARCHAR, max_length=100, is_partition_key=True),
]
schema = CollectionSchema(fields=fields)
```

Then we declare a collection, set the number of partitions:
`collection = Collection(name="xxx", schema=schema, num_partitions=500)`
num_partitions=500 means we will create 500 partitions inside this collection. The names of these partitions are internally defined as "_default_0", "_default_1" ...... In fact, we don't need these names.

When we insert data, we assign a user name for each record:
```
collection.insert({"id": 1, "vector": [0.12, 0.23...], "username": "Mike"})
collection.insert({"id": 2, "vector": [0.56, 0.01...], "username": "Bob"})
......
```

The records are automatically distributed to different partitions, according to the hash value of the username. "Mike" maybe assign to partition_25, "Bob" maybe assign to partition_8, etc. Some user names might share the same partition since two hash values could be equal.

When we search, we use filtering expression to search the records of a user:
`collection.search(expr="username=='Mike'", ......)`

With username=='Mike', milvus can quickly know which partition to be searched according to the hash value of "Mike". Then it only searches the records of "Mike" inside the partition_25. Sometimes there are some other usernames in partition_25, no matter, milvus will skip them.
 
We can imagine the collection is a building. Each partition is a room. The building has 500 rooms.
If we have 1000 users, we invite them to different rooms, according to their names. Each room has 2 users ideally. Actually, some rooms might have 3 users, some rooms might be empty.

If there are 1 million users,  each room has 2000 users ideally.
When we want to find a user, from his name we can quickly know in which room he is. There are some other users in this room, but no matter, we can ignore them.

**Comment by jubingc** [comment]
Thanks for the response about the collection/partition. Can I know the relationship between "database" and "collection/partition"? 

My goal is to provision one database for one customer. Customers have the freedom to create collections and partitions as they need. I don't manage how users use their collections/partitions. Apart from [managing database](https://milvus.io/docs/v2.3.x/manage_databases.md), I would like to learn more about the performance management of databases. How is search performance ensured for different databases?

**Reply by yhmo** [reply]
The maximum number of database is 64.
A database has multiple collections, a collection has multiple partitions.
I think search performance has nothing to do with database.

**Comment by jubingc** [comment]
Another question about the collection count. According to [document](https://milvus.io/docs/limit_collection_counts.md), the number of general capacity is the product of (collections, shards, partitions). In your example, when you assign 500 partitions to a collection. You end up with 500 collections. These 500 collections consume the limit of `maxGeneralCapacity`. Is my understanding correct?

When you said

> milvus will add more threads to manage the collection/partition.

Are these threads (goroutines?) created on coordinators or query/data nodes?


When you mentioned 

>  it is strongly recommended to keep the number of collections/partitions under 1000(depends on resources(CPU/RAM, high-perf disk), with rich resources, it is better to keep the number under 5000).

Does it still hold regardless of the size of the cluster? If I add more query nodes to the cluster, is the recommended number of collections/partitions increased?

**Reply by yhmo** [reply]
The maxGeneralCapacity default value is 65536. In fact, when you create 10000+ collections in the cluster, you will see the cluster response slowly or run into a weird state, no response, crash, etc.  As I mentioned, it is better to keep the number of collections/partitions under 5000. I mean the total number of collections/partitions. For example, you have 100 collections, each collection has 100 partitions, total number is 100 * 100 = 10000, the milvus server might run into trouble.

Each collection/partition will add more goroutines to querycoord/querynode/datacoord/datanode.

No matter how many querynodes/datanodes in the cluster, the total number of collections/partitions should be controlled under 5000.

**Reply by mihailyanchev** [reply]
Quick question here. When you say the total number of collections/partitions needs to be kept under 5000, do you mean the ones loaded in memory or in total? Also, do you mean in cluster mode or in the standalone mode?

I'm asking because I implemented a sharding strategy where I keep the number of collections fixed, but allocate data separated by business logic in dedicated partitions. I load partitions in memory only on demand and unload all data from memory in a nightly pipeline job. So now I am wondering if my whole premise was wrong.

**Reply by xiaofan-luan** [reply]
> Quick question here. When you say the total number of collections/partitions needs to be kept under 5000, do you mean the ones loaded in memory or in total? Also, do you mean in cluster mode or in the standalone mode?
> 
> I'm asking because I implemented a sharding strategy where I keep the number of collections fixed, but allocate data separated by business logic in dedicated partitions. I load partitions in memory only on demand and unload all data from memory in a nightly pipeline job. So now I am wondering if my whole premise was wrong.

You can definitely manually load and release colleciton/partiton based on your needs.

Be sure you are using 2.5.4 since we made many progress on serving more collections in one cluster. With latest 2.5.4 you should be able to serve more than 10K collections and 100K partitions in total.

**Reply by mihailyanchev** [reply]
Thank you for the quick reply! This sounds very good. Just to confirm these limits are for collections/partitions loaded in-memory, correct?

**Reply by xiaofan-luan** [reply]
Not really. I kind of based on the blockers.
But reduce the number of collections loaded definitely help

**Reply by mihailyanchev** [reply]
Thanks! So is scaling the cluster vertically a solution to this? I would assume that up to a point.
Or does one need to separate the data across clusters in order to avoid these limits and scale horizontally?

EDIT:
Follow up question. If you have a lot of collections/partitions, but they are not in-memory is the RAM still the resource bottleneck?

**Reply by yhmo** [reply]
> Thanks! So is scaling the cluster vertically a solution to this? I would assume that up to a point. Or does one need to separate the data across clusters in order to avoid these limits and scale horizontally?
> 
> EDIT: Follow up question. If you have a lot of collections/partitions, but they are not in-memory is the RAM still the resource bottleneck?

I don't think it is a good solution to separate the data across clusters. We need to optimize milvus to handle more collections/partitions.

If you have a lot of collections/partitions, but you didn't call load() to load them, they are not in memory. But if you want to query/search them, you must call load() to load them into memory.
